File: Editorial/82_Remove_Duplicates_from_Sorted_List_II/Go/1_Sentinel_Head_Predecessor.go
func deleteDuplicates(head *ListNode) *ListNode {
    // sentinel
    sentinel := &ListNode{0, head}
    // predecessor = the last node
    // before the sublist of duplicates
    pred := sentinel
    for head != nil {
        // If it's a beginning of duplicates sublist
        // skip all duplicates
        if head.Next != nil && head.Val == head.Next.Val {
            // move till the end of duplicates sublist
            for head.Next != nil && head.Val == head.Next.Val {
                head = head.Next
            }
            // Skip all duplicates
            pred.Next = head.Next
            // otherwise, move predecessor
        } else {
            pred = pred.Next
        }
        // move forward
        head = head.Next
    }
    return sentinel.Next
}

File: Editorial/0000-0100/0082_Remove_Duplicates_from_Sorted_List_II/TypeScript/1_Sentinel_Head_Predecessor.ts
function deleteDuplicates(head: ListNode | null): ListNode | null {
    // sentinel
    let sentinel: ListNode = new ListNode(0, head);
    // predecessor = the last node
    // before the sublist of duplicates
    let pred: ListNode = sentinel;
    while (head !== null) {
        // If it's a beginning of duplicates sublist
        // skip all duplicates
        if (head.next !== null && head.val === head.next.val) {
            // move till the end of duplicates sublist
            while (head.next !== null && head.val === head.next.val) {
                head = head.next;
            }
            // Skip all duplicates
            pred.next = head.next;
            // otherwise, move predecessor
        } else {
            pred = pred.next;
        }
        // move forward
        head = head.next;
    }
    return sentinel.next;
}
